### PR TITLE
Advanced motion

### PR DESCRIPTION
**Pull request by CodingKoopa**
This PR adds:
- Proper encoder and gyrometer support.
- Odometry support.
- A command for following a trajectory.
- Drivetrain simulation support.

This contains several commits from #10, as I wanted mode commands for this. Perhaps review that first, and once it's merged the overlapping commits here will no longer show up.

### CODE REVIEW

**Comment by CodingKoopa** [comment]
Okay, so here's what's going to go on with all of this. This PR still contains the code for trajectory following, and all of its prerequisites. To prevent this PR from getting too large, though, I am cutting it off there.

I have the following things saved:
- Vision Subsystem code: Code for a vision subsystem, for tracking targets. I am thinking I will put this in the hands of @matthewrshane, for his own PR.
- Intake Subsystem code: Code for ball tracking for the intake subsystem. Not currently working.
- Galactic Search code: Code for galactic search, including following and tracking the power cells. Depends on the vision subsystem. I'm not sure how I'm going to handle this. I can't totally yield control of this, perhaps I will join it with the aforementioned PR.

With that, I am leaving this PR up for review now.

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -69,15 +71,166 @@
     public static final int kPortControllerManip = 1;
   }
 
-  /** Constants for the drivetrain subsystem. */
+  /**
+   * Constants for the drivetrain subsystem.
+   *
+   * <p>These values are susceptible to unit errors, so unit comments are required for all of these.
+   */
   public static final class DrivetrainConstants {
```
Comment (RyanStyron) [inline src/main/java/frc/robot/Constants.java:79]: Were all of these verified, particularly `kMaxTrajectoryVelocity` and `kMaxTrajectoryAcceleration`?

Comment (CodingKoopa) [reply src/main/java/frc/robot/Constants.java:79]: - The speed multiplier constants are fine.
- `kMaxTrajectoryVelocity` and `kMaxTrajectoryAcceleration`, I have no idea. I think they are going to be okay as a starting point.
- All of the physical robot constants are fine.
- The encoder constants are fine.
- The angular feedforward constants are wrong. We have calibrated these, but that information is on the unpushed Robot2021Char commit.
- The feedback constants should be fine as a starting point.

Not everything is verified, but there's not much more we can do right now. I think we can keep it for now, and fix the angular feedforwards later.

Comment (RyanStyron) [reply src/main/java/frc/robot/Constants.java:79]: Ok, all else looks good, so I'll merge.

**Pull request merged**